### PR TITLE
ethminer: provide a CUDA-free version

### PR DESCRIPTION
**Pull request by Atemu**
###### Motivation for this change

The cuda version is broken and my AMDGPU system doesn't need it to be built with CUDA.

cc @nand0p @noneucat @r-burns @danieldk 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 120588` at 12675a78 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>ethminer</li>
    <li>ethminer-cuda</li>
  </ul>
</details>
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>ethminer-free</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl2` is a deprecated license, please check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/tools/misc/ethminer/default.nix:83:5:
```
   |
83 |     license = licenses.gpl2;
   |     ^
```
</li>
  </ul>
</details>

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -3470,7 +3470,9 @@ in
 
   ethash = callPackage ../development/libraries/ethash { };
 
-  ethminer = callPackage ../tools/misc/ethminer { };
+  ethminer = ethminer-cuda;
+  ethminer-cuda = callPackage ../tools/misc/ethminer { };
+  ethminer-free = callPackage ../tools/misc/ethminer { cudatoolkit = null; withCuda = false; };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:3475]:
```suggestion
  ethminer-free = callPackage ../tools/misc/ethminer { withCuda = false; };
```

File: pkgs/tools/misc/ethminer/default.nix [hunk 2]
```diff
@@ -8,29 +9,30 @@
   boost,
   makeWrapper,
   cudatoolkit,
+  withCuda ? cudatoolkit != null,
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/ethminer/default.nix:12]:
```suggestion
  withCuda ? true,
```

File: pkgs/tools/misc/ethminer/default.nix [hunk 3]
```diff
@@ -8,29 +9,30 @@
...
   mesa,
   ethash,
   opencl-info,
   ocl-icd,
   openssl,
   pkg-config,
   cli11
-}:
+}@args:
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/ethminer/default.nix:20]: Why do we need args here?

File: pkgs/tools/misc/ethminer/default.nix [hunk 4]
```diff
@@ -1,5 +1,6 @@
 {
   lib,
+  stdenv,
   clangStdenv,
   fetchFromGitHub,
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/ethminer/default.nix:5]:
```suggestion
{ lib
, stdenv
, clangStdenv
, fetchFromGitHub
```

Please format the inputs in this style.

File: pkgs/tools/misc/ethminer/default.nix (hunk 3 above)
Comment (Atemu) [reply pkgs/tools/misc/ethminer/default.nix:20]: Infinite recursion in let binding later.

File: pkgs/tools/misc/ethminer/default.nix (hunk 4 above)
Comment (Atemu) [reply pkgs/tools/misc/ethminer/default.nix:5]: Could we leave style nitpickings on portions I didn't even really touch out of this? Thanks.

If you want to "fix" code style (we don't even have a style guide) feel free to make PRs specific to that purpose (I'd genuinely like that!) but please don't ask me to do that for you in PRs which are not.  
I get it, style is important. I try to "fix" sections which I perceive as not conforming to the "norms" when there's a good opportunity (i.e. I'm reworking it in large parts anyways) but you cannot expect people to go out of their way to do that sort of thing where it wouldn't be necessary.

I didn't open ethminer's default.nix and said to myself "Boy golly, I'm going to fix this file's styling!", I opened it to make a functional change with actual value to users.

As a contributor, I see such reviews as entirely unnecessary and even discouraging.

File: pkgs/tools/misc/ethminer/default.nix (hunk 2 above)
Comment (Atemu) [reply pkgs/tools/misc/ethminer/default.nix:12]: I'm not sure what the benefit of that would be?

My way makes it a bit more explicit why cudatoolkit (unfree) is needed. When calling the package from a Nixpkgs which doesn't have unfree packages in it for example, it'd default to a free build which is a useful property I think.

Comment (danieldk) [reply pkgs/tools/misc/ethminer/default.nix:12]: This is really badly documented, but the default name for this flag now seems to be `cudaSupport, with its value set to `config.cudaSupport`:

https://github.com/NixOS/nixpkgs/blob/350f9bd822e1c5017482e6364dbafda36ff4157a/pkgs/top-level/all-packages.nix#L1723

This has the benefit that if someone already has `cudaSupport = true; allowUnfree = true` in their nixpkgs `config`, CUDA support would also be enabled automatically for ethminer.

**Review by danieldk — COMMENTED** [review]
It's nice to have have derivations buildable without CUDA support as well. LGTM. I added one small suggestion to use `cudaSupport`, since it is more standardized in nixpkgs.

**Comment by danieldk** [comment]
@ofborg build ethminer ethminer-cuda ethminer-free

**Review by danieldk — APPROVED** [review]

**Pull request merged**